components/client-pangolin-s2s/src/feemarket.rs

```diff
@@ -106,4 +111,67 @@ impl RelayStrategy for PangolinRelayStrategy {
         }
         false
     }
+
+    async fn callback<
+        P: MessageLane,
+        SourceClient: MessageLaneSourceClient<P>,
+        TargetClient: MessageLaneTargetClient<P>,
+    >(
+        &self,
+        reference: &RelayReference<P, SourceClient, TargetClient>,
+        decide: bool,
+    ) {
+        let mut total_cost = P::SourceChainBalance::zero();
+
+        let confirmation_transaction_cost = reference
+            .lane_source_client
+            .estimate_confirmation_transaction()
+            .await;
+
+        let delivery_transaction_cost = match reference
+            .lane_target_client
+            .estimate_delivery_transaction_in_source_tokens(
+                reference.hard_selected_begin_nonce
+                    ..=(reference.hard_selected_begin_nonce + reference.index as MessageNonce),
+                reference.selected_prepaid_nonces,
+                reference.selected_unpaid_weight,
+                reference.selected_size as u32,
+            )
+            .await
+        {
+            Ok(v) => v,
+            Err(err) => {
+                log::debug!(
+                    target: "bridge",
+                    "Failed to estimate delivery transaction cost: {:?}. No nonces selected for delivery",
+                    err,
+                );
+                P::SourceChainBalance::zero()
+            }
+        };
+
+        total_cost = total_cost.saturating_add(confirmation_transaction_cost);
+        total_cost = total_cost.saturating_add(delivery_transaction_cost);
+        let data = CalculateData {
+            decide,
+            total_cost: total_cost.unique_saturated_into(),
+        };
+        self.callback.handle(data).await;
```

Comment:
5. Callback, The data includes
    - `decide`: whether to relay this transaction
    - `total_cost`: all the cost of the relay transaction